### PR TITLE
Albert huggingface

### PR DESCRIPTION
**Pull request by tomkoren21**
add support to Albert/HuggingFace

### CODE REVIEW

File: onnx2kerastl/elementwise_layers.py [hunk 1]
```diff
@@ -136,7 +137,10 @@ def convert_elementwise_sub(node, params, layers, lambda_func, node_name, keras_
 
     except (IndexError, ValueError):
         logger.warning('Failed to use keras.layers.Subtract. Fallback to TF lambda.')
-        layers[node_name] = input_0 - input_1
+        if input_0_is_np and not input_1_is_np:  # constant - tensor does not parse well
+            layers[node_name] = - (input_1 - input_0)
```
Comment (doronHarnoy) [inline onnx2kerastl/elementwise_layers.py:141]: 🥲

File: onnx2kerastl/linear_layers.py [hunk 2]
```diff
@@ -31,24 +33,26 @@ def convert_gemm(node, params, layers, lambda_func, node_name, keras_name):
...
         logger.debug('Transposing W matrix.')
         keras_weights[0] = keras_weights[0].transpose()
-
     # Estimate input/output neurons
-    input_channels, output_channels = keras_weights[0].shape
+    input_channels, output_channels = keras_weights[0].shape[-2:]
     logger.debug('Input units %s, output units %s.', input_channels, output_channels)
+    if len(keras_weights[0].shape) > 3: #N-dim tensor multipication Dense doesn't work
+        assert len(node.input) == 2
+        layers[node_name] = tf.matmul(layers[node.input[0]], layers[node.input[1]], name=keras_name)
```
Comment (doronHarnoy) [inline onnx2kerastl/linear_layers.py:41]: As far as i know Dense layer can perform N-dim multipications

File: onnx2kerastl/operation_layers.py [hunk 3]
```diff
@@ -414,3 +407,8 @@ def target_layer(x, axis=axis, keepdims=keepdims):
     lambda_layer = keras.layers.Lambda(target_layer, name=keras_name)
     layers[node_name] = lambda_layer(input_0)
     lambda_func[keras_name] = target_layer
+
+
+def convert_reciprocal(node, params, layers, lambda_func, node_name, keras_name):
+    input_0 = ensure_tf_type(layers[node.input[0]],  name="%s_const" % keras_name)
+    layers[node_name] = tf.math.reciprocal(input_0)
```
Comment (doronHarnoy) [inline onnx2kerastl/operation_layers.py:414]: This operation is converted to TFOpLambda also?

File: onnx2kerastl/operation_layers.py [hunk 4]
```diff
@@ -158,8 +156,7 @@ def handle_axis_change(target_data_format: str, config: Dict[str, Any]) -> Dict[
 
         return handle_axis_change
 
-    reduce_mean_layer.get_special_data_format_handler = get_special_data_format_handler
-    layers[node_name] = reduce_mean_layer(input_0)
+    layers[node_name] = K.mean(input_0, axis=axes, keepdims=keepdims)
```
Comment (doronHarnoy) [inline onnx2kerastl/operation_layers.py:159]: I think that we need this layer to create its special_data_format_handler when the layer is being converted from last to first.
How ` K.mean` is being converted from last to first now? (its axis should change accordingly)

File: onnx2kerastl/operation_layers.py [hunk 5]
```diff
@@ -223,9 +220,8 @@ def convert_sqrt(node, params, layers, lambda_func, node_name, keras_name):
         assert AttributeError('More than 1 input for sqrt layer.')
 
     input_0 = ensure_tf_type(layers[node.input[0]], name="%s_const" % keras_name)
+    layers[node_name] = K.sqrt(input_0)
 
-    sqrt_layer = OnnxSqrt(name=keras_name)
-    layers[node_name] = sqrt_layer(input_0)
```
Comment (doronHarnoy) [inline onnx2kerastl/operation_layers.py:228]: I would keep `OnnxSqrt` just because it looks better than another TFOPLambda and we already made the effort to create it

File: onnx2kerastl/pooling_layers.py [hunk 6]
```diff
@@ -154,14 +154,16 @@ def convert_global_avg_pool(node, params, layers, lambda_func, node_name, keras_
...
     :return: None
     """
-
     input_0 = ensure_tf_type(layers[node.input[0]], layers[list(layers)[0]], name="%s_const" % keras_name)
-
-    global_pool = keras.layers.GlobalAveragePooling2D(data_format='channels_first', name=keras_name)
+    tensor_dim = len(input_0.shape)
+    if tensor_dim == 3:
+        global_pool = keras.layers.GlobalAveragePooling1D(data_format='channels_first', name=keras_name)
+    else:
```
Comment (doronHarnoy) [inline onnx2kerastl/pooling_layers.py:161]: I Recommend adding here `if tensor_dim == 4:` for this condition and another `else` after that throws an Exception explaining what happened

File: onnx2kerastl/reshape_layers.py [hunk 7]
```diff
@@ -89,8 +89,19 @@ def convert_gather(node, params, layers, lambda_func, node_name, keras_name):
...
+            indices = layers[node.input[1]]
+        else:
+            indices = layers[node.input[1]].tolist()
+        if "is_embedding" in params:
+            if len(input_0.shape) == 2:
+                emb = tf.keras.layers.Embedding(input_0.shape[0], input_0.shape[1], weights=[layers[node.input[0]]],
+                                                name=keras_name)
+                layers[node_name] = emb(indices)
+            else:
+                raise AttributeError("Cannot transform gather into embedding with non 2D array")
```
Comment (doronHarnoy) [inline onnx2kerastl/reshape_layers.py:102]: Lets think if this else block will not throw on unnecessary times also as Gather op can be a from of embedding but also can be just a Gather from a tensor

File: onnx2kerastl/reshape_layers.py [hunk 8]
```diff
@@ -193,8 +204,11 @@ def target_layer(x):
                     flatten = keras.layers.Flatten(name=keras_name)
                     layers[node_name] = flatten(input_0)
                 else:
-                    reshape = keras.layers.Reshape(np.int32(input_1[1:]), name=keras_name)
-                    layers[node_name] = reshape(input_0)
+                    if abs(len(input_0.shape) - len(input_1)) == 1:  # keras reshape don't work
+                        layers[node_name] = tf.reshape(input_0, input_1, name=keras_name)
```
Comment (doronHarnoy) [inline onnx2kerastl/reshape_layers.py:208]: Can you describe in what cases this if block will be used?

File: pyproject.toml [hunk 9]
```diff
@@ -12,6 +12,8 @@ tensorflow-macos = {version = "2.8.0", markers = "platform_machine  == 'arm64'"}
 onnx = "^1.11.0"
 keras-data-format-converter = "0.0.17"
 tensorflow-addons = "^0.17.1"
+transformers = {extras = ["onnx"], version = "^4.20.1"}
+sentencepiece = "^0.1.96"
```
Comment (doronHarnoy) [inline pyproject.toml:16]: Onnx is currently not supported on Mac's M1 chip, In order to support this package on m1 we are putting on onnx dependency the following (you will see it after rebase):
markers = "platform_machine  == 'x86_64'"

Lets check if the two packages you are adding here `transformers` `sentencepiece` can be run on m1 before we add them or put on them the marker to be used only on intel chip

**Review by doronHarnoy — APPROVED** [review]
LGTM, did the models tests run?

File: onnx2kerastl/converter.py [hunk 10]
```diff
@@ -183,6 +183,8 @@ def onnx_to_keras(onnx_model, input_names,
 
                 if node_input in weights:
                     logger.debug('Found in weights, add as a numpy constant.')
+                    if node_type == "Gather":
+                        node_params['is_embedding'] = True
```
[the remark this answers is not in the document]
Comment (tomkoren21) [reply onnx2kerastl/converter.py:187]: It iterates on weights

File: onnx2kerastl/operation_layers.py (hunk 3 above)
Comment (tomkoren21) [reply onnx2kerastl/operation_layers.py:414]: converted to 1/x

File: onnx2kerastl/linear_layers.py (hunk 2 above)
Comment (tomkoren21) [reply onnx2kerastl/linear_layers.py:41]: Things is we multiple Nd matrices here (no weights) - will change the condition from 3 to 2

File: onnx2kerastl/operation_layers.py (hunk 4 above)
Comment (tomkoren21) [reply onnx2kerastl/operation_layers.py:159]: reverting to previous version

File: onnx2kerastl/operation_layers.py (hunk 5 above)
Comment (tomkoren21) [reply onnx2kerastl/operation_layers.py:228]: OK

File: onnx2kerastl/pooling_layers.py (hunk 6 above)
Comment (tomkoren21) [reply onnx2kerastl/pooling_layers.py:161]: OK

File: onnx2kerastl/reshape_layers.py (hunk 7 above)
Comment (tomkoren21) [reply onnx2kerastl/reshape_layers.py:102]: Agreed to throw an error for now in unexpected input

File: onnx2kerastl/reshape_layers.py (hunk 8 above)
Comment (tomkoren21) [reply onnx2kerastl/reshape_layers.py:208]: Changed this to a new condition that considers batch changes only

**Comment by tomkoren21** [comment]
Locally - deeplab/googlenet/mobilenetv2/shufflenet/vit fails in both master and this branch
